### PR TITLE
[FLINK-30282] Fix Logical type ROW lost inner field's nullability after converting to RelDataType

### PR DESCRIPTION
**Pull request by LadyForest**
## What is the purpose of the change

This pull request fixes the issue that Flink SQL's logical type `Row` lost the inner field's nullability after converting to `RelDataType`. E.g. `a Row<f0 INT NOT NULL, f1 STRING>` is converted to `a Row<f0 INT, f1 STRING>`

## Brief change log
-  Let`ExtendedSqlRowTypeNameSpec` pass the inner field's nullability when deriving type.
- Specify the `StructKind.PEEK_FIELDS_NO_EXPAND` to create a struct type. This is required because in `FlinkTypeFactory#createTypeWithNullability`, only `PEEK_FIELDS_NO_EXPAND` preserves the inner field's nullability.

## Verifying this change

This change is already covered by existing tests `SqlToOperationConverterTest#testCreateTableWithFullDataTypes`. The nullability check has been tweaked. However, the issue that the precision of the `TIME` and `Row` type's inner comment loss still exists, so the `TODO` is not removed. Beyond this, the `table.q` file also covers the change check.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduces a new feature? no
  - If yes, how is the feature documented? not applicable

### CODE REVIEW

**Comment by LadyForest** [comment]
@flinkbot run azure

**Comment by LadyForest** [comment]
@flinkbot run azure

**Comment by LadyForest** [comment]
@fsk119 would you spare some time to take a look?

**Comment by snuyanzin** [comment]
@LadyForest this change looks reasonable however now there are some conflicts...
Do you mind if I ask you to rebase?

**Comment by LadyForest** [comment]
> @LadyForest this change looks reasonable however now there are some conflicts... Do you mind if I ask you to rebase?

Sure. I'll do the rebase.

**Comment by LadyForest** [comment]
Hi @snuyanzin, could you mind taking a look?

File: flink-table/flink-sql-client/src/test/resources/sql/table.q [hunk 1]
```diff
@@ -964,7 +998,7 @@ CREATE TABLE IF NOT EXISTS daily_orders (
  PRIMARY KEY(dt, `user`) NOT ENFORCED
 ) PARTITIONED BY (dt) WITH (
  'connector' = 'filesystem',
- 'path' = '$VAR_BATCH_PATH',
+ 'path' = '/var/folders/xd/9dp1y4vd3h56kjkvdk426l500000gp/T/junit5781726749537647921/7bf278b4-9932-4e24-af7b-1126df4667405030013249695173150',
```
Comment (snuyanzin) [inline flink-table/flink-sql-client/src/test/resources/sql/table.q:1001]: this i didn't get...
where does this magic name come from?

Comment (LadyForest) [reply flink-table/flink-sql-client/src/test/resources/sql/table.q:1001]: > this i didn't get... can you please elaborate where does this magic name come from?

Good catch! `$VAR_BATCH_PATH` is defined in `CliClientITCase` and serves as a placeholder to be replaced at runtime. I overwrote this by mistake and will revert this.

**Comment by LadyForest** [comment]
One concern is that for a nullable row type, does it make sense to let the inner field not be null?
In other words, should we not support `ROW<a INT NOT NULL, b STRING, c DOUBLE NOT NULL>`? 
WDYT @snuyanzin

**Comment by LadyForest** [comment]
> According to SQL standards, a record's value (with structure type) is considered null if and only if all of its fields are null.
https://issues.apache.org/jira/browse/CALCITE-3627

**Pull request closed without merging**